test/widgets/content_test.dart

```diff
@@ -92,17 +92,75 @@ void main() {
 
   TestZulipBinding.ensureInitialized();
 
-  Future<void> prepareContentBare(WidgetTester tester, String html) async {
-    await tester.pumpWidget(Builder(
-      builder: (context) {
-        return MaterialApp(
+  /// Parses the given [html] and pumps the corresponding content widgets.
+  ///
+  /// Use the minimum "extra" dependencies necessary (like
+  /// [wrapWithPerAccountStoreWidget]), and for each one used, explain why
+  /// in a code comment. See existing callers for examples.
+  ///
+  /// When testing content widgets, prefer this function (adapting it as needed)
+  /// instead of writing ad-hoc content-preparing code. It helps us track our
+  /// content widgets' dependencies.
+  // TODO(#488) For content that we need to show outside a per-message or
+  //   per-account context, make sure to include tests that don't provide
+  //   such context.
+  Future<List<Route<dynamic>>> prepareContentBare(WidgetTester tester, String html, {
+    bool wrapWithScaffold = false,
+    bool wrapWithPerAccountStoreWidget = false,
+    bool wrapWithMessageContent = false,
+    TextStyle? defaultTextStyleOverride,
+  }) async {
+    Widget widget = BlockContentList(nodes: parseContent(html).nodes);
+
+    if (wrapWithMessageContent) {
+      widget = MessageContent(
+        message: eg.streamMessage(content: html),
+        content: parseContent(html));
+    }
+
+    if (defaultTextStyleOverride != null) {
+      widget = DefaultTextStyle(style: defaultTextStyleOverride,
+        child: widget);
+    }
+
+    if (wrapWithScaffold) {
+      widget = Scaffold(body: widget);
+    }
+
+    if (wrapWithPerAccountStoreWidget) {
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+      widget = PerAccountStoreWidget(accountId: eg.selfAccount.id, child: widget);
+    }
+
+    widget = GlobalStoreWidget(child: widget);
+    addTearDown(testBinding.reset);
+
+    prepareBoringImageHttpClient();
+
+    final pushedRoutes = <Route<dynamic>>[];
+    final testNavObserver = TestNavigatorObserver()
+      ..onPushed = (route, prevRoute) => pushedRoutes.add(route);
```

Comment:
This particular feature makes me uneasy to have here in the generic shared helper because there's no easy way to search for the particular test cases that rely on it. One has to look for call sites that use the return value, and potentially trace that data flow onward.

The industrial-strength way to solve that would be to refactor this helper from a function `prepareContentBare` into being a class instead. Then `pushedRoutes` would be data on that class — either a public field, or a private one with a method like `takePushedRoutes` to consume it — and one could look for references to the relevant field or method.

But rather than undertake that, I think a good solution would be to just make this feature a bit more generic, with just enough of a hook for the caller to be able to integrate in the needed way with the `MaterialApp`: have this prepare method take something like `List<NavigatorObserver> navObservers = const []`. Then the three call sites stay a few lines longer than in this version, but I think that's OK. Those lines are a tad verbose, but aren't all wasted either — they provide some explicitness which is nice.

If we wanted to simplify the `TestNavigatorObserver` / `pushedRoutes` stuff in those call sites, a good direction would be to extend `TestNavigatorObserver` itself. It's a class in our own test code; we use it in a number of other test files too, and it looks like in every case we do the exact same thing with it.